libraries/Wire/src/utility/twi.cpp

```diff
@@ -541,26 +541,19 @@ void i2c_attach_slave_tx_callback(i2c_t *obj, void (*function)(void*), void* pWi
  * @param length Number of bytes to read
  * @return status
  */
-i2c_status_enum i2c_slave_write_buffer(i2c_t *obj, uint8_t *data, uint16_t length)
-{
+i2c_status_enum i2c_slave_write_buffer(i2c_t *obj, uint8_t *data, uint16_t length){
     struct i2c_s *obj_s = I2C_S(obj);
-    uint8_t i = 0;
-    i2c_status_enum ret = I2C_OK;
+    if (    (obj_s->tx_count + length) > obj->tx_rx_buffer_size) 
+        return I2C_DATA_TOO_LONG;
 
-    if (length > obj->tx_rx_buffer_size) {
-        ret = I2C_DATA_TOO_LONG;
-    } else {
-        /* check the communication status */
-        for (i = 0; i < length; i++) {
-            *obj_s->tx_buffer_ptr++ = *(data + i);
-        }
-        obj_s->tx_count = length;
-        obj_s->tx_buffer_ptr = obj_s->tx_buffer_ptr - length;
+    uint8_t i = 0;
+    for (i; i < length; i++) {
+        *obj_s->tx_buffer_ptr++ = *(data + i);
+        obj_s->tx_count += length;
```

Comment:
Sadly that's not correct. That would increment `tx_count` by `length` in every iteration of this `for` loop, doing an increase by length-squared. This has to be pulled out of the `for` loop`.